### PR TITLE
Fix `std` to work with `numeric_only` for `pandas` 2.0

### PR DESCRIPTION
**Pull request by j-bennet**
This fixes the test failure in upstream:

```
dask/dataframe/tests/test_arithmetics_reduction.py::test_datetime_std_across_axis1_null_results[False]: TypeError: float() argument must be a string or a real number, not 'Timestamp'
dask/dataframe/tests/test_arithmetics_reduction.py::test_datetime_std_across_axis1_null_results[True]: TypeError: float() argument must be a string or a real number, not 'Timestamp'
```

Unlike https://github.com/dask/dask/pull/9952, this PR does not make `numeric_only` changes anywhere in DataFrame/Series methods, besides `std`.

Xref https://github.com/dask/dask/issues/9736.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by j-bennet** [comment]
Tests are now passing.

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -1609,13 +1610,19 @@ def test_dataframe_quantile(method, expected, numeric_only):
 
         assert (result == expected[1]).all().all()
 
-        with assert_numeric_only_default_warning(numeric_only):
+        with warnings.catch_warnings(record=True):
+            warnings.filterwarnings("ignore", category=FutureWarning)
+            # pandas issues a warning with 1.5, but not 1.3
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_dataframe.py:1615]: Hmm, based on the comment here I don't quite understand this change as shouldn't be expecting a warning in 1.3

https://github.com/dask/dask/blob/e893d5333ffca172d4c0abc46c3058055c436b82/dask/dataframe/tests/test_dataframe.py#L1522-L1530

Maybe I'm missing something tough

File: dask/dataframe/tests/test_dataframe.py [hunk 2]
```diff
@@ -1609,13 +1610,19 @@ def test_dataframe_quantile(method, expected, numeric_only):
...
             expected = df.quantile(axis=1, **numeric_only_kwarg)
         with assert_numeric_only_default_warning(numeric_only):
             result = ddf.quantile(axis=1, method=method, **numeric_only_kwarg)
-        assert_eq(result, expected)
 
-        with pytest.raises(ValueError), check_numeric_only_deprecation():
+        with check_numeric_only_deprecation():
+            assert_eq(result, expected)
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_dataframe.py:1621]: It looks like we're now warning at compute time -- is there a way we can avoid that?

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -124,6 +124,49 @@ def wrapper(self, *args, **kwargs):
...
+    """Update numeric_only to get rid of no_default, and possibly warn about default value.
+    TODO: should move to numeric_only decorator. See https://github.com/dask/dask/pull/9952
+    """
+    if is_dataframe_like(df):
+        warn_numeric_only = False
+        if numeric_only is no_default:
+            if PANDAS_GT_200:
+                numeric_only = False
+            else:
+                warn_numeric_only = True
```
[the remark this answers is not in the document]
Comment (j-bennet) [reply dask/dataframe/core.py:152]: Yes, the point of returning `{}` here is to make sure we never pass `no_default` further to `pandas`, to use their default instead.

File: dask/dataframe/core.py [hunk 4]
```diff
@@ -2498,7 +2540,7 @@ def std(
                 skipna=skipna,
                 ddof=ddof,
                 enforce_metadata=False,
-                numeric_only=numeric_only,
+                **numeric_kwargs,
```
[the remark this answers is not in the document]
Comment (j-bennet) [reply dask/dataframe/core.py:2543]: Merged.

File: dask/dataframe/tests/test_arithmetics_reduction.py [hunk 5]
```diff
@@ -1544,21 +1545,35 @@ def test_datetime_std_creates_copy_cols(axis):
 
     ddf = dd.from_pandas(pdf, 3)
 
+    kwargs = {} if numeric_only is None else {"numeric_only": numeric_only}
+
+    ctx = contextlib.nullcontext()
+    if numeric_only is False or (PANDAS_GT_200 and numeric_only is None):
```
[the remark this answers is not in the document]
Comment (j-bennet) [reply dask/dataframe/tests/test_arithmetics_reduction.py:1551]: Yeah. It's a little different in different places, but most can be generalized.

File: dask/dataframe/tests/test_arithmetics_reduction.py [hunk 6]
```diff
@@ -1630,13 +1653,30 @@ def test_datetime_std_across_axis1_null_results(skipna):
...
+
+    pctx = contextlib.nullcontext()
+    dctx = contextlib.nullcontext()
+    if numeric_only is False or (PANDAS_GT_200 and numeric_only is None):
+        dctx = pytest.raises(NotImplementedError, match="numeric_only")
+        pctx = pytest.raises(TypeError)
+    elif numeric_only is None:
+        dctx = pytest.warns(FutureWarning, match="numeric_only")
+        if PANDAS_GT_130:
+            pctx = pytest.warns(FutureWarning, match="numeric_only")
```
[the remark this answers is not in the document]
Comment (j-bennet) [reply dask/dataframe/tests/test_arithmetics_reduction.py:1667]: Right now, in both `quantile` and `std`, I'm using the same helper function, `_numeric_only_maybe_warn`. It replaces the old `@_numeric_only` decorator that the rest of the methods use.

If you look here, I'm not checking for a lower threshhold of pandas version to issue a warning, just that we're below 2.0:

https://github.com/dask/dask/blob/db4761aefc04305b9f723262eb9d78948fec7d60/dask/dataframe/core.py#L152

Pandas has different warning behavior for these two functions below 1.5, one of them warns, the other doesn't. I think it actually makes sense to warn of future changes to numeric behavior with both. And this way, I can also reuse the helper function.

File: dask/dataframe/tests/test_dataframe.py (hunk 1 above)
Comment (j-bennet) [reply dask/dataframe/tests/test_dataframe.py:1615]: Yes, I made Dask warn below 1.5 here:

https://github.com/dask/dask/blob/db4761aefc04305b9f723262eb9d78948fec7d60/dask/dataframe/core.py#L152

See [this comment](https://github.com/dask/dask/pull/9960/files#r1109023514) for the reasoning.

**Review by jrbourbeau — APPROVED** [review]
Thanks @j-bennet

**Pull request merged**